vortex-ipc/src/codecs/message_reader/monoio.rs

```diff
@@ -0,0 +1,180 @@
+use crate::flatbuffers::ipc::Message;
+
+struct MonoIoMessageReader<R: AsyncReadRent + Unpin> {
+    // TODO(ngates): swap this for our own mutable aligned buffer so we can support direct reads.
+    read: R,
+    message: BytesMut,
+    prev_message: BytesMut,
+    finished: bool,
+}
+
+impl<R: AsyncReadRent + Unpin> MonoIoMessageReader<R> {
+    pub async fn try_new(read: R) -> VortexResult<Self> {
+        let mut reader = Self {
+            read,
+            message: BytesMut::new(),
+            prev_message: BytesMut::new(),
+            finished: false,
+        };
+        reader.load_next_message().await?;
+        Ok(reader)
+    }
+
+    async fn load_next_message(&mut self) -> VortexResult<bool> {
+        // FIXME(ngates): how do we read into a stack allocated thing?
+        let len_buf = self.read.read_exact_into(Vec::with_capacity(4)).await?;
+
+        let len = u32::from_le_bytes(len_buf.as_slice().try_into()?);
+        if len == u32::MAX {
+            // Marker for no more messages.
+            return Ok(false);
+        }
+
+        // TODO(ngates): we may be able to use self.message.split() and then swap back after.
+
+        let message = self
+            .read
+            .read_exact_into(BytesMut::with_capacity(len as usize))
+            .await?;
+
+        // Validate that the message is valid a flatbuffer.
+        let _ = root::<Message>(message.as_ref())?;
+
+        self.message = message;
+
+        Ok(true)
+    }
+}
+
+impl<R: AsyncReadRent + Unpin> MessageReader for MonoIoMessageReader<R> {
+    fn peek(&self) -> Option<Message> {
+        if self.finished {
+            return None;
+        }
+        // The message has been validated by the next() call.
+        Some(unsafe { root_unchecked::<Message>(&self.message) })
+    }
+
+    async fn next(&mut self) -> VortexResult<Message> {
+        if self.finished {
+            panic!("StreamMessageReader is finished - should've checked peek!");
+        }
+        self.prev_message = self.message.split();
+        if !self.load_next_message().await? {
+            self.finished = true;
+        }
+        Ok(unsafe { root_unchecked::<Message>(&self.prev_message) })
+    }
+
+    async fn buffers(&mut self) -> VortexResult<Vec<Buffer>> {
+        let Some(chunk_msg) = unsafe { root_unchecked::<Message>(&self.message) }.header_as_chunk()
+        else {
+            // We could return an error here?
+            return Ok(Vec::new());
+        };
+
+        // Read all the column's buffers
+        let mut offset = 0;
+        let mut buffers = Vec::with_capacity(chunk_msg.buffers().unwrap_or_default().len());
+        for buffer in chunk_msg.buffers().unwrap_or_default().iter() {
+            let _skip = buffer.offset() - offset;
+            self.read.skip(buffer.offset() - offset).await?;
+
+            // TODO(ngates): read into a single buffer, then Arc::clone and slice
+            let bytes = BytesMut::zeroed(buffer.length() as usize);
```

Comment:
I think this is prime usecase for readv. Right now we read everything but once you start projecting and reading parts then readv would be ideal. If you take the code from my branch you will have the logic to avoid skipping

Reply:
I've ported monoio over to readv, but not futures since they don't have `read_vectored_exact`. It might be interesting anyway to see the performance difference.